### PR TITLE
feat(rules): templating actions

### PR DESCRIPTION
**Pull request by UnderKoen**
Relates to #500  and #3378 

https://github.com/user-attachments/assets/809ee622-af8a-4151-9293-8ce4e9dbdb91

### CODE REVIEW

**Comment by youngcw** [comment]
Could you put in some more examples of what is possible with this?  I see that you can pull in field values, can you pull in things like account balance?

**Comment by UnderKoen** [comment]
You can currently only pull in information about the transaction. This is limited to anything what is currently possible with data available to rules. So not the name of the payee or name of the category. I saw an todo `// TODO: Add matches op support for payees, accounts, categories.` which woudl probaly also make this possible!

The main use case for this would currently be cleaning imported transactions. Or altering existing transactions.

Or changing the date of new income to the first of the month.
`{{floor (add (year date) (div (month date) 12))}}-{{add (mod (month date) 12) 1}}-1`

Is this now very verbose but we could add another helper for adding duration to dates.

The helpers I currently implemented are:
- `regex value regex replacement`
- `add num1 num2 ...`
- `sub num1 num2 ...`
- `div num1 num2 ...`
- `mul num1 num2 ...`
- `mod num1 num2 ...`
- `floor num`
- `ceil num`
- `round num`
- `abs num`
- `min num1 num2 ...`
- `max num1 num2 ...`
- `fixed num1 num2`
- `day date`
- `month date`
- `year date`
- `format date format`

Values available:
- today (date)
- account (id)
- amount (number)
- cleared (boolean)
- date (date)
- error (?)
- imported_id (id)
- imported_payee (string)
- is_child (boolean)
- is_parent (boolean)
- notes (string)
- parent_id (id)
- payee (id)
- reconciled (boolean)
- schedule (id)
- sort_order (number)
- starting_balance_flag (boolean)
- tombstone (boolean)
- transfer_id (id)

This list can be extended by adding more values at `rules.ts#569` like how `today` is added.

**Comment by youngcw** [comment]
There probably needs to be a syntax check in the rule edit window since its so easy to have bad syntax with something like this.

Are these supposed to run when manually creating a transaction?  My testing doesn't seem to work in that case.  I think its working for file imports though.

Im worried that there could be some issues with how the csv import checks for duplicates if someone has a templated rule.  That will require more testing.

**Comment by UnderKoen** [comment]
@youngcw 
> There probably needs to be a syntax check in the rule edit window since its so easy to have bad syntax with something like this.
Will add

> Are these supposed to run when manually creating a transaction? My testing doesn't seem to work in that case. I think its working for file imports though.
I just tested this and it has been working but not consistently, I think the rules are not always run?

> Maybe we should look at moving the rules inside of the check that the csv import does for duplication matching.
Seems smart, I wouldn't have a clue how to achieve this. Could this be fixxed in an separatem PR and making this feature expiremental?

**Comment by youngcw** [comment]
Making this experimental would probably be good for now.

**Comment by youngcw** [comment]
> * tombstone (boolean)

Can this be set? or is it just available to use?

**Comment by youngcw** [comment]
> Out of scope of this PR I think but wouldn't this be best solved with field like `imported_payee`? Just have `imported_amount`, `imported_notes`, etc.

I don't think that will fix if the amount gets changed

**Comment by UnderKoen** [comment]
> Im worried that there could be some issues with how the csv import checks for duplicates if someone has a templated rule. That will require more testing.

> edit: So the csv importer can't see the matching transaction if there is a big change (change the amount sign for example), but the rule gets run before the real deduplication goes. Maybe we should look at moving the rules inside of the check that the csv import does for duplication matching.

I don't think I understand what the intented behaviour is and how to check the behaviour?

**Comment by youngcw** [comment]
> > Im worried that there could be some issues with how the csv import checks for duplicates if someone has a templated rule. That will require more testing.
> 
> > edit: So the csv importer can't see the matching transaction if there is a big change (change the amount sign for example), but the rule gets run before the real deduplication goes. Maybe we should look at moving the rules inside of the check that the csv import does for duplication matching.
> 
> I don't think I understand what the intented behaviour is and how to check the behaviour?

If you import a csv and then import the same csv again you should see that all the "new" transactions get matched with the existing transaction and shown to you in the import window to decide what to do.  So compare that to if you have a rule that modifies the amount using a template rule and there wont be a match even though its the same transaction and will get deduplicated after the rules run.

**Comment by UnderKoen** [comment]
Does this not work with the exported CSV's? I have no rules active and use the demo file
image

**Comment by Djurq** [comment]
This would be awesome!

**Comment by youngcw** [comment]
with the CSV?  Its not that the CSV import isn't working.  The CSV importer shows which transactions are new and which ones match an existing transaction.  If the rules change the amounts, for example, then  the matching doesn't work since the transaction has changed too much.

**Comment by UnderKoen** [comment]
@youngcw How does this look, when I import an CSV I don't see anywhere that it matches even with no rules active. After importing with and without rules there are no duplicates.

**Comment by youngcw** [comment]
Import the same csv twice and you will see extra lines in the import window the second time.  Those lines will show the matching

**Comment by UnderKoen** [comment]
@youngcw I don't see any extra line?

2024-09-10 19 46 chrome_HV526F7C2X
2024-09-10 19 46 chrome_PVxZgABVCy

**Comment by youngcw** [comment]
Hmm. Something must have broke.  Im not seeing it anymore either.

**Comment by youngcw** [comment]
https://deploy-preview-3179.demo.actualbudget.org/  This is the deploy for v24.8.0.  It still works with the matching so you can see what it is supposed to be doing.

**Comment by youngcw** [comment]
Its possible that this would work fine if the table wasn't messed up prior to this.

**Comment by UnderKoen** [comment]
Ahha, thanks I really tough I was going crazy

**Comment by UnderKoen** [comment]
@youngcw it seems that the templating rules work with the CSV import. Could you confirm?

File: packages/loot-core/src/server/accounts/rules.test.ts [hunk 1]
```diff
@@ -316,6 +316,88 @@
...
+        test(template, () => {
+          const action = new Action('set', 'notes', '', { template });
+          const item = { notes: 'Sarah Condition' };
+          action.exec(item);
+          expect(item.notes).toBe(expected);
+        });
+      }
+
+      testHelper('{{regex notes "/[aeuio]/g" "a"}}', 'Sarah Candataan');
+      testHelper('{{regex notes "/[aeuio]/" ""}}', 'Srah Condition');
```
[the remark this answers is not in the document]
Comment (UnderKoen) [reply packages/loot-core/src/server/accounts/rules.test.ts:341]: There is no global flag so it should only replace the first letter

**Comment by youngcw** [comment]
Ill need to look more, but the csv matching logic seems to be working.  The only thing that is a little weird is that the line that shows the existing transaction that matches will show the after rules version and the line that shows the incoming version doesn't have rules applied, but they show as matching.  I guess that makes sense, but maybe we could do something to make it clear why they are different in the future.

**Comment by youngcw** [comment]
Also, this will need a pretty thorough docs page since this gets complicated fast.

File: packages/desktop-client/src/components/settings/Experimental.tsx [hunk 2]
```diff
@@ -110,6 +110,9 @@ export function ExperimentalFeatures() {
             >
               <Trans>Customizable reports page (dashboards)</Trans>
             </FeatureToggle>
+            <FeatureToggle flag="actionTemplating">
```
Comment (matt-fidd) [inline packages/desktop-client/src/components/settings/Experimental.tsx:113]:
```suggestion
            <FeatureToggle
              flag="actionTemplating"
              feedbackLink="https://github.com/actualbudget/actual/issues/3606"
            >
```

**Review by matt-fidd — CHANGES_REQUESTED** [review]
Thanks for this!

Just a quick couple of changes with how the settings toggle is presented

**Comment by matt-fidd** [comment]
Opened tracking issue #3606 for experimental feedback

**Review by matt-fidd — APPROVED** [review]
Amazing, thank you for this! I'd like to get it merged in as is for now to release it into the hands of `edge` users and gather some feedback before the next release cycle.

**Review by youngcw — APPROVED** [review]

**Pull request merged**